client-spark/common/src/main/java/org/apache/spark/shuffle/ShuffleHandleInfo.java

```diff
@@ -85,33 +126,125 @@ public int getShuffleId() {
     return shuffleId;
   }
 
-  public boolean isExistingFaultyServer(String serverId) {
-    return faultyServerReplacements.containsKey(serverId);
+  @VisibleForTesting
+  protected boolean isMarkedAsFaultyServer(String serverId) {
+    return faultyServerToReplacements.containsKey(serverId);
   }
 
-  public ShuffleServerInfo useExistingReassignmentForMultiPartitions(
-      Set<Integer> partitionIds, String faultyServerId) {
-    return createNewReassignmentForMultiPartitions(partitionIds, faultyServerId, null);
+  public Set<ShuffleServerInfo> getExistingReplacements(String faultyServerId) {
+    return faultyServerToReplacements.get(faultyServerId);
   }
 
-  public ShuffleServerInfo createNewReassignmentForMultiPartitions(
-      Set<Integer> partitionIds, String faultyServerId, ShuffleServerInfo replacement) {
-    if (replacement != null) {
-      faultyServerReplacements.put(faultyServerId, replacement);
+  public void updateReassignment(
+      Set<Integer> partitionIds, String faultyServerId, Set<ShuffleServerInfo> replacements) {
+    if (replacements == null) {
+      return;
     }
 
-    replacement = faultyServerReplacements.get(faultyServerId);
+    faultyServerToReplacements.put(faultyServerId, replacements);
+    // todo: optimize the multiple for performance
     for (Integer partitionId : partitionIds) {
-      List<ShuffleServerInfo> replicaServers = partitionToServers.get(partitionId);
-      for (int i = 0; i < replicaServers.size(); i++) {
-        if (replicaServers.get(i).getId().equals(faultyServerId)) {
-          Map<Integer, ShuffleServerInfo> replicaReplacements =
-              failoverPartitionServers.computeIfAbsent(
-                  partitionId, k -> JavaUtils.newConcurrentMap());
-          replicaReplacements.put(i, replacement);
+      Map<Integer, List<ShuffleServerInfo>> replicaServers =
+          partitionReplicaAssignedServers.get(partitionId);
+      for (Map.Entry<Integer, List<ShuffleServerInfo>> serverEntry : replicaServers.entrySet()) {
+        List<ShuffleServerInfo> servers = serverEntry.getValue();
+        if (servers.stream()
+            .map(x -> x.getId())
+            .collect(Collectors.toSet())
+            .contains(faultyServerId)) {
+          Set<ShuffleServerInfo> tempSet = new HashSet<>();
+          tempSet.addAll(replacements);
+          tempSet.removeAll(servers);
+          servers.addAll(tempSet);
```

Comment:
This will keep the faulty servers in `servers`, is this expected?

Reply:
Yes.

The name of `faulty servers` here is not accurate, (But here I will not change this, maybe later.) because due to tight memory, the NO_BUFFER will be threw. For these cases, we'd better to reassign partition server to write data.

So the reserving these server is to fetch partial data when reading.

Reply:
Why don't we read from all servers until all blocks are read? Then we would not need to add replacement server on failure here?

Reply:
Oh,  I think I need to say the reassign occurs on the writer side. Sorry I still don’t catch your thought.